[14.0][IMP] auditlog: prevent cascading delete of logs when models or fields are unlinked

StefanRijnhart opened the pull request:
When a field or a model is unlinked, keep the related audit logs. Denormalize the field and model info on the logs and log lines so that the information is still available after the deletion of the related data model.

Also, to improve the performance of the deletion of fields and models, add indexes on the log's `model_id` and log line's `field_id`.

Port of https://github.com/OCA/server-tools/pull/1262.

PieterPaulussen commented on auditlog/models/log.py:
```diff
@@ -9,7 +10,11 @@ class AuditlogLog(models.Model):
+        "ir.model", string="Model", index=True, ondelete="set null"
+    )
+    model_name = fields.Char(string=u"Model Name", readonly=True)
+    model_model = fields.Char(string=u"Technical Model Name", readonly=True)
```
@StefanRijnhart the `unicode` is probably not required here?

StefanRijnhart replied:
Ah yes. As a matter of fact it wasn't necessary in Python 2.7, but they are particularly offensive in Python 3.x. Removed now.

rconjour commented on auditlog/migrations/14.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,73 @@
+    )
+    cr.execute(
+        """
+    ALTER TABLE auditlog_log
```
The alignment of those lines seems odd

StefanRijnhart replied:
Same as https://github.com/OCA/server-tools/blob/8.0/auditlog/migrations/8.0.1.4.0/pre-migration.py. I don't think there is a leading convention on the formatting of multiline SQL statements so I would prefer to leave it like this.

StefanRijnhart replied:
The only change was applied to the position of the quotes, by python-black in the prehook, and we can't derive from that.

rconjour reviewed (changes requested):
Looking good, only the syntax offset in the migration seems odd.

Commit:
[IMP] auditlog: prevent cascading delete of logs when models or fields are unlinked

When a field or a model is unlinked, keep the related audit logs. Denormalize
the field and model info on the logs and log lines so that the information
is still available after the deletion of the related data model.

Also, to improve the performance of the deletion of fields and models,
add indexes on the log's model_id and log line's field_id.

Co-Authored-By: Stefan Rijnhart <stefan@opener.am>

StefanRijnhart commented:
Oh, just pushed a version that set the new fields to readonly in the view.

PieterPaulussen commented on auditlog/models/log.py:
```diff
@@ -9,7 +10,11 @@ class AuditlogLog(models.Model):
+        "ir.model", string="Model", index=True, ondelete="set null"
+    )
+    model_name = fields.Char(readonly=True)
+    model_model = fields.Char(string="Technical Model Name", readonly=True)
```
Stupid question probably, but why aren't these stored related fields?

StefanRijnhart replied:
Pieter, you *wrote* this code ;-) The reason they are not stored related fields is because those would get wiped when the field/model is deleted, whereas the exact point of these fields is to keep the info when the field/model is deleted.

StefanRijnhart replied:
This OK for you @PieterPaulussen?

PieterPaulussen commented on auditlog/models/log.py:
```diff
@@ -35,5 +60,30 @@ class AuditlogLogLine(models.Model):
-    field_name = fields.Char("Technical name", related="field_id.name")
-    field_description = fields.Char("Description", related="field_id.field_description")
+    field_name = fields.Char("Technical name", readonly=True)
+    field_description = fields.Char("Description", readonly=True)
```
Same question as above?

dreispt commented:
/ocabot merge nobump
